src/offset.rs

```diff
@@ -490,6 +491,104 @@ where
+            .next()
+            .and_then(|start: OffsetItem| self.offsets.peek().map(|end: &OffsetItem| (start, *end)))
+            .map(|(start, end)| {
+                let slot_start: usize = start.try_into().expect("offset value should be in range");
+                let slot_end: usize = end.try_into().expect("offset value should be in range");
+                let slot_length = slot_end
+                    .checked_sub(slot_start)
+                    .expect("offsets should be monotonically increasing");
+
+                let taken = self.data.by_ref().take(slot_length).collect::<Vec<_>>();
+
+                debug_assert!(
+                    taken.len() == slot_length,
+                    "underlying data array does not have enough elements, \
+                    expected {} elements for this slot, found {}",
+                    slot_length,
+                    taken.len()
+                );
+
+                taken
+            })
+    }
+}
+
+impl<T, OffsetItem: OffsetElement, Buffer: BufferType> IntoIterator
+    for Offset<T, false, OffsetItem, Buffer>
+where
+    T: IntoIterator,
+    <Buffer as BufferType>::Buffer<OffsetItem>: IntoIterator<Item = OffsetItem>,
+    OffsetIntoIter<T, OffsetItem, Buffer>: Iterator<Item = Vec<<T as IntoIterator>::Item>>,
+{
+    type Item = Vec<<T as IntoIterator>::Item>;
+    type IntoIter = OffsetIntoIter<T, OffsetItem, Buffer>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        OffsetIntoIter {
+            data: self.data.into_iter(),
+            offsets: self.offsets.into_iter().peekable(),
+        }
+    }
+}
+
+impl<T, OffsetItem: OffsetElement, Buffer: BufferType> IntoIterator
+    for Offset<T, true, OffsetItem, Buffer>
+where
+    T: IntoIterator,
+    Bitmap<Buffer>: IntoIterator<Item = bool>,
+    <Buffer as BufferType>::Buffer<OffsetItem>: IntoIterator<Item = OffsetItem>,
+    OffsetIntoIter<T, OffsetItem, Buffer>: Iterator<Item = Vec<<T as IntoIterator>::Item>>,
+{
+    type Item = Option<Vec<<T as IntoIterator>::Item>>;
+    type IntoIter = Map<
+        Zip<
+            <Bitmap<Buffer> as IntoIterator>::IntoIter,
+            <OffsetIntoIter<T, OffsetItem, Buffer> as IntoIterator>::IntoIter,
+        >,
+        fn((bool, Vec<<T as IntoIterator>::Item>)) -> Self::Item,
+    >;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.offsets
+            .validity
+            .into_iter()
+            // We use zip insead of map because a null value may have a
+            // positive slot length. That is, a null value may occupy a
+            // non-empty memory space in the data buffer.
+            // Thus the underlying iterator must be advanced even for null
+            // values as indicated by the validity bitmap.
+            .zip(OffsetIntoIter {
+                data: self.data.into_iter(),
+                offsets: self.offsets.data.into_iter().peekable(),
+            })
+            .map(|(validity, value): (bool, Vec<_>)| validity.then_some(value))
```

Comment:
Maybe you should add a comment that we need the eager evaluation of `then_some` here?

Reply:
Could you elaborate a bit? I'm not sure if I fully understand? Since the underlying iterator must always be advanced.

Reply:
Oh right, the `data` iterator is advanced by `OffsetIntoIter`. We should maybe consider special-casing this to avoid the allocation of the `Vec` for `null` values using something like https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.advance_by when "skipping" over a null element. But that's an optimization we can do in a follow-up PR.